vernac/comHints.ml

```diff
@@ -99,20 +98,41 @@ let interp_hints ~poly h =
       let gr = Smartlocate.global_with_alias c in
       (PathHints [gr], hint_globref gr)
     | HintsConstr c ->
-      let () = warn_deprecated_hint_constr () in
       let env = Global.env () in
       let sigma = Evd.from_env env in
       let c, uctx = Constrintern.interp_constr env sigma c in
       let subst, uctx = UState.normalize_variables uctx in
       let c = EConstr.Vars.subst_univs_constr subst c in
       let diff = UState.context_set uctx in
       let c =
-        if poly then (c, Some diff)
+        if hint_proxy () then
+          let sigma = Evd.merge_context_set UState.UnivRigid sigma diff in
+          let avoid = Environ.ids_of_named_context_val (Environ.named_context_val env) in
+          (* This is a hack to ensure that including two modules defining anonymous
+            hints will not trigger an already defined error. *)
+          let name = Printf.sprintf "Private_hint_local_%x_" (ModPath.hash @@ Global.current_modpath ()) in
+          let name = Namegen.next_global_ident_away (Id.of_string name) avoid in
+          let ctx = Evd.univ_entry ~poly sigma in
+          let c = EConstr.to_constr sigma c in
+          let cb =
+            Declare.(DefinitionEntry (definition_entry ~univs:ctx ~opaque:false c))
+          in
+          let c =
+            Declare.declare_constant ~local:Locality.ImportNeedQualified ~name
+              ~kind:Decls.(IsDefinition Definition)
+              cb
```

Comment:
This function is scheduled for removal from `Declare`, what would the higher-level API need to be used here?